File: tool/uapicodegen/uapicodegen/uapi/internal/generation/GenerateRandomSelect.py
from typing import Dict, List, Union

from ...internal.generation.GenerateContext import GenerateContext


def generate_random_select(possible_selects: dict[str, object], ctx: 'GenerateContext') -> object:
    possible_select = possible_selects[ctx.fn_scope]
    return sub_select(possible_select, ctx)


def sub_select(possible_select_section: object, ctx: 'GenerateContext') -> object:
    if isinstance(possible_select_section, list):
        selected_field_names = []

        for field_name in possible_select_section:
            if ctx.random_generator.next_boolean():
                selected_field_names.append(field_name)

        return selected_field_names
    elif isinstance(possible_select_section, dict):
        selected_section: dict[str, object] = {}

        for key, value in possible_select_section.items():
            if ctx.random_generator.next_boolean():
                result = sub_select(value, ctx)
                if isinstance(result, dict) and not result:
                    continue
                selected_section[key] = result

        return selected_section
    else:
        raise ValueError('Invalid possible_select_section')

File: lib/java/src/main/java/io/github/telepact/internal/generation/GenerateRandomSelect.java
//|
//|  Copyright The Telepact Authors
//|  SPDX-License-Identifier: Apache-2.0
//|

package io.github.telepact.internal.generation;

import java.util.*;

public class GenerateRandomSelect {

    public static Object generateRandomSelect(Map<String, Object> possibleSelects, GenerateContext ctx) {
        var possibleSelect = possibleSelects.get(ctx.fnScope);
        return subSelect(possibleSelect, ctx);
    }

    private static Object subSelect(Object possibleSelectSection, GenerateContext ctx) {
        if (possibleSelectSection instanceof List) {
            List<String> selectedFieldNames = new ArrayList<>();

            for (Object fieldName : (List<?>) possibleSelectSection) {
                if (ctx.randomGenerator.nextBoolean()) {
                    selectedFieldNames.add((String) fieldName);
                }
            }

            return selectedFieldNames.stream().sorted().toList();
        } else if (possibleSelectSection instanceof Map) {
            final var m = (Map<String, Object>) possibleSelectSection;
            Map<String, Object> selectedSection = new HashMap<>();

            List<String> keys = m.keySet().stream().sorted().toList();

            for (String key : keys) {
                final var value = m.get(key);
                if (ctx.randomGenerator.nextBoolean()) {
                    Object result = subSelect(value, ctx);
                    if (result instanceof Map && ((Map<?, ?>) result).isEmpty()) {
                        continue;
                    }
                    selectedSection.put(key, result);
                }
            }

            return selectedSection;
        } else {
            throw new IllegalArgumentException("Invalid possibleSelectSection");
        }
    }
}
